### PR TITLE
[GLUTEN-1648][VL] Add max_by/min_by aggregate function support

### PR DESCRIPTION
**Pull request by Yohahaha**

### CODE REVIEW

**Comment by Yohahaha** [comment]
Spark's max_by has a little different with Presto/Velox in implementation, detailed discussion here https://github.com/facebookincubator/velox/pull/4971 .

Seems Spark UT does not specify these corner cases, let's add these functions for now, and will be keep modify above PR if it is necessary.

**Comment by Yohahaha** [comment]
@rui-mo @PHILO-HE could you help take a look?

**Comment by PHILO-HE** [comment]
Hi @Yohahaha, thanks for your patch! Recently, to relieve the burden of velox code rebase, we prefer directly contributing velox code to upstream (exceptions are some fixes on the code only exists in oap/velox, etc.). Could you please try to get the velox patch merged to upstream?
It's OK to submit this Gluten PR for early code review & Gluten CI verifications. I will add `pending velox rebase` tag. Thanks!

**Comment by rui-mo** [comment]
> Spark's max_by has a little different with Presto/Velox in implementation, detailed discussion here [facebookincubator/velox#4971](https://github.com/facebookincubator/velox/pull/4971) .
> 
> Seems Spark UT does not specify these corner cases, let's add these functions for now, and will be keeping modify above PR if it is necessary.

Hi @Yohahaha, how large would the difference affect the computing? There could be some risks if the functions are offloaded to Velox but with incorrect result.

**Comment by Yohahaha** [comment]
> Hi @Yohahaha, thanks for your patch! Recently, to relieve the burden of velox code rebase, we prefer directly contributing velox code to its upstream (exceptions are some fixes on the code only exists in oap/velox, etc.). Could you please try to get the velox patch merged to its upstream? It's OK to submit this Gluten PR for early code review & Gluten CI verifications. I will add `pending velox rebase` tag. Thanks!

This PR does not include codes which need push to upstream, please check related oap/velox link.

**Comment by Yohahaha** [comment]
> > Spark's max_by has a little different with Presto/Velox in implementation, detailed discussion here [facebookincubator/velox#4971](https://github.com/facebookincubator/velox/pull/4971) .
> > Seems Spark UT does not specify these corner cases, let's add these functions for now, and will be keeping modify above PR if it is necessary.
> 
> Hi @Yohahaha, how large would the difference affect the computing? There could be some risks if the functions are offloaded to Velox but with incorrect result.

It's a minor change, different behaviors see below
```
SELECT max_by(x, y) FROM VALUES ('a', 10), ('b', 50), ('c', 50) AS tab(x, y)

Spark result is c
Presto result is b
```

Spark does not specify `max_by/min_by` as deterministic, so I pending update above PR.

**Comment by rui-mo** [comment]
> SELECT max_by(x, y) FROM VALUES ('a', 10), ('b', 50), ('c', 50) AS tab(x, y)
> 
> Spark result is c
> Presto result is b

@Yohahaha Thank you, I see. So the results can be different when there are equal values, right? 

> Spark does not specify `max_by/min_by` as deterministic, so I pending update above PR.

Could you explain this statement more a bit? Does that mean Spark's behavior is also random?

**Comment by Yohahaha** [comment]
> So the results can be different when there are equal values, right?
Yes.

> Could you explain this statement more a bit? Does that mean Spark's behavior is random?
If we have 2 equal values and both are largest of group, Spark will return second, however Presto return first. Since input data in agg is non-deterministic, Velox community think there is no need to distinguish that difference.

**Comment by rui-mo** [comment]
@Yohahaha Could you help open an issue for this corner case in Gluten?

**Comment by Yohahaha** [comment]
Convert to draft and wait align behavior with Spark.

**Comment by PHILO-HE** [comment]
Please note we have moved the code related to substrait from oap/velox into gluten. So for your proposed [changes](https://github.com/Yohahaha/velox/commit/30bf9026dc8ab99afc7da4574514dc39350a9f55) in oap/velox, please apply them in `gluten/cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc`. Thanks!

**Comment by Yohahaha** [comment]
Wait oap/velox rebase 8.2

**Comment by PHILO-HE** [comment]
Hi @Yohahaha, as oap/velox is aligned with the upstream, this PR can be merged. Could you please do a rebase?

**Comment by Yohahaha** [comment]
@rui-mo I see this [change](https://github.com/oap-project/velox/blame/b8fc63ec35f049a49e417ba4b0845c2b1e367748/velox/functions/sparksql/aggregates/AverageAggregate.cpp#L365-L368) and may_by/min_by support also need change here, this is your expected?

**Comment by rui-mo** [comment]
> @rui-mo I see this [change](https://github.com/oap-project/velox/blame/b8fc63ec35f049a49e417ba4b0845c2b1e367748/velox/functions/sparksql/aggregates/AverageAggregate.cpp#L365-L368) and may_by/min_by support also need change here, this is your expected?

@Yohahaha This change is to allow companion function registration through config. What do you need for max_by and min_by?

**Comment by Yohahaha** [comment]
> > @rui-mo I see this [change](https://github.com/oap-project/velox/blame/b8fc63ec35f049a49e417ba4b0845c2b1e367748/velox/functions/sparksql/aggregates/AverageAggregate.cpp#L365-L368) and may_by/min_by support also need change here, this is your expected?
> 
> @Yohahaha This change is to allow companion function registration through config. What do you need for max_by and min_by?

Spark's min_by/max_by need overwrite presto's, but current register function does not provide these two bool args.

https://github.com/oap-project/velox/blob/b8fc63ec35f049a49e417ba4b0845c2b1e367748/velox/functions/sparksql/aggregates/Register.cpp#L33C33-L43

**Comment by rui-mo** [comment]
@Yohahaha This is my draft PR https://github.com/facebookincubator/velox/pull/7110. I can add these two bool args to min_by and max_by if needed.

**Comment by Yohahaha** [comment]
https://github.com/oap-project/gluten/issues/3472

**Comment by rui-mo** [comment]
In the long term, I think we will only register spark sql functions after all the needed functions are supported. Currently, presto sql functions are registered just because the lack of functionality of spark sql functions.

**Comment by rui-mo** [comment]
We've added the overwrite config for max_by and mix_by at [Register.cpp#L41](https://github.com/oap-pproject/velox/blob/update/velox/functions/sparksql/aggregates/Register.cpp#L41). Maybe this PR can be firstly enabled with that. cc @PHILO-HE

**Comment by Yohahaha** [comment]
> We've added the overwrite config for max_by and mix_by at [Register.cpp#L41](https://github.com/oap-project/velox/blob/update/velox/functions/sparksql/aggregates/Register.cpp#L41). Maybe this PR can be firstly enabled with that. cc @PHILO-HE

Thanks, I will try later.

**Comment by Yohahaha** [comment]
> We've added the overwrite config for max_by and mix_by at [Register.cpp#L41](https://github.com/oap-project/velox/blob/update/velox/functions/sparksql/aggregates/Register.cpp#L41). Maybe this PR can be firstly enabled with that. cc @PHILO-HE

depends on https://github.com/oap-project/velox/pull/426

**Comment by rui-mo** [comment]
@Yohahaha There are code style failures. Please refer to https://github.com/oap-project/gluten/blob/main/docs/developers/NewToGluten.md#javascala-code-style for code style adjustment.

**Comment by Yohahaha** [comment]
CI passed.

**Comment by Yohahaha** [comment]
please help review again, thanks! @rui-mo @PHILO-HE

**Review by PHILO-HE — DISMISSED** [review]
Thanks!

**Comment by rui-mo** [comment]
@Yohahaha Could you resolve the conflicts? Thanks.

**Review by PHILO-HE — APPROVED** [review]

File: backends-clickhouse/src/main/scala/io/glutenproject/utils/CHExpressionUtil.scala [hunk 1]
```diff
@@ -153,6 +153,8 @@ case class EncodeDecodeValidator() extends FunctionValidator {
 object CHExpressionUtil {
 
   final val CH_AGGREGATE_FUNC_BLACKLIST: Map[String, FunctionValidator] = Map(
+    MAX_BY -> DefaultValidator(),
```
Comment (PHILO-HE) [inline backends-clickhouse/src/main/scala/io/glutenproject/utils/CHExpressionUtil.scala:156]: cc @zzcclp

Comment (zzcclp) [reply backends-clickhouse/src/main/scala/io/glutenproject/utils/CHExpressionUtil.scala:156]: LGTM

**Pull request merged**